### ISSUE TITLE
[SDK-49] [Android] [expo-dev-client] Crash on android API_22

### ISSUE THREAD

**Issue by GabrielDierks** [expo/expo#24896]
### Minimal reproducible example

https://github.com/GabrielDierks/test-expo-android-22/

### Summary

When creating a new expo app with the expo-dev-client on the latest SDK 49 on android SDK 22 the app crashes.

## Steps:

1. Create new expo project
2. Add the `expo-dev-client` project
3. Install all dependencies and run `npx expo prebuild --clean && npx expo run:android -d`
4. Build succeeds!
5. App crashes on open, adb logcat log:
`E/SoLoader(23418): couldn't find DSO to load: libflipper.so caused by: dlopen failed: cannot locate symbol "BN_set_flags" referenced by "/data/app/com.gabrielkadmos.testyyyy-1/lib/arm64/libflipper.so"... result: 0
E/AndroidRuntime(23418): FATAL EXCEPTION: FlipperConnectionThread
E/AndroidRuntime(23418): Process: com.gabrielkadmos.testyyyy, PID: 23418
E/AndroidRuntime(23418): java.lang.NoClassDefFoundError: <clinit> failed for class com.facebook.flipper.android.EventBase; see exception in other thread
E/AndroidRuntime(23418): 	at com.facebook.flipper.android.FlipperThread.run(FlipperThread.java:25)
E/AndroidRuntime(23418): FATAL EXCEPTION: FlipperEventBaseThread
E/AndroidRuntime(23418): Process: com.gabrielkadmos.testyyyy, PID: 23418
E/AndroidRuntime(23418): java.lang.UnsatisfiedLinkError: couldn't find DSO to load: libflipper.so caused by: dlopen failed: cannot locate symbol "BN_set_flags" referenced by "/data/app/com.gabrielkadmos.testyyyy-1/lib/arm64/libflipper.so"... result: 0`

6. Build config plugin to disable Flipper (see example repo ./application-plugin.js)
7. App crashes again with the following logs
`E/AndroidRuntime(23733): FATAL EXCEPTION: main
E/AndroidRuntime(23733): Process: com.gabrielkadmos.testyyyy, PID: 23733
E/AndroidRuntime(23733): java.lang.RuntimeException: Unable to start activity ComponentInfo{com.gabrielkadmos.testyyyy/com.gabrielkadmos.testyyyy.MainActivity}: java.lang.NoSuchFieldException: accessFlags
E/AndroidRuntime(23733): 	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2325)
E/AndroidRuntime(23733): 	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2387)
E/AndroidRuntime(23733): 	at android.app.ActivityThread.access$800(ActivityThread.java:151)
E/AndroidRuntime(23733): 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1303)
E/AndroidRuntime(23733): 	at android.os.Handler.dispatchMessage(Handler.java:102)
E/AndroidRuntime(23733): 	at android.os.Looper.loop(Looper.java:135)
E/AndroidRuntime(23733): 	at android.app.ActivityThread.main(ActivityThread.java:5254)
E/AndroidRuntime(23733): 	at java.lang.reflect.Method.invoke(Native Method)
E/AndroidRuntime(23733): 	at java.lang.reflect.Method.invoke(Method.java:372)
E/AndroidRuntime(23733): 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:903)
E/AndroidRuntime(23733): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:698)
E/AndroidRuntime(23733): Caused by: java.lang.NoSuchFieldException: accessFlags
E/AndroidRuntime(23733): 	at java.lang.Class.getDeclaredField(Class.java:890)
E/AndroidRuntime(23733): 	at expo.modules.ReactActivityDelegateWrapper.onCreate(ReactActivityDelegateWrapper.kt:126)
E/AndroidRuntime(23733): 	at com.facebook.react.ReactActivity.onCreate(ReactActivity.java:46)
E/AndroidRuntime(23733): 	at com.gabrielkadmos.testyyyy.MainActivity.onCreate(MainActivity.java:20)
E/AndroidRuntime(23733): 	at android.app.Activity.performCreate(Activity.java:5990)
E/AndroidRuntime(23733): 	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1106)
E/AndroidRuntime(23733): 	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2278)
E/AndroidRuntime(23733): 	... 10 more
W/ActivityManager(19092):   Force finishing activity 1 com.gabrielkadmos.testyyyy/expo.modules.devlauncher.launcher.errors.DevLauncherErrorActivity`

### Environment

  expo-env-info 1.0.5 environment info:
    System:
      OS: macOS 14.0
      Shell: 5.9 - /bin/zsh
    Binaries:
      Node: 16.17.0 - ~/.nvm/versions/node/v16.17.0/bin/node
      Yarn: 1.22.18 - /usr/local/bin/yarn
      npm: 8.15.0 - ~/.nvm/versions/node/v16.17.0/bin/npm
    Managers:
      CocoaPods: 1.11.3 - /opt/homebrew/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 23.0, iOS 17.0, macOS 14.0, tvOS 17.0, watchOS 10.0
      Android SDK:
        API Levels: 30, 31, 32, 33, 34
        Build Tools: 29.0.2, 30.0.2, 30.0.3, 31.0.0, 32.0.0, 32.1.0, 33.0.0, 33.0.0, 34.0.0
        System Images: android-22 | Google APIs ARM 64 v8a, android-32 | Google APIs ARM 64 v8a, android-32 | Google Play ARM 64 v8a, android-33 | Google APIs ARM 64 v8a, android-34 | Google APIs ARM 64 v8a
    IDEs:
      Android Studio: 2022.2 AI-222.4459.24.2221.10121639
      Xcode: 15.0/15A240d - /usr/bin/xcodebuild
    npmPackages:
      expo: ~49.0.13 => 49.0.13
      react: 18.2.0 => 18.2.0
      react-native: 0.72.5 => 0.72.5
    npmGlobalPackages:
      eas-cli: 5.4.0
      expo-cli: 6.3.10
    Expo Workflow: managed

**Comment by GabrielDierks** [comment]
Any update on this? We cannot debug our app on API Level < 26 with the dev-client at the moment which is quite severe!

**Comment by ide** [comment]
@gabrieldonadel Does expo-dev-client require API Level 26+? Or is there a way to gracefully disable functionality and communicate the issue with a notice if attempting to access `accessFlags` throws a `NoSuchFieldException` instead of crashing the app?

**Comment by gabrieldonadel** [comment]
> @gabrieldonadel Does expo-dev-client require API Level 26+? Or is there a way to gracefully disable functionality and communicate the issue with a notice if attempting to access `accessFlags` throws a `NoSuchFieldException` instead of crashing the app?

From my understanding, the problem is this line https://github.com/expo/expo/blob/7e01ae7e9802c7ea7853b8d6f3163dbb36cbccac/packages/expo/android/src/main/java/expo/modules/ReactActivityDelegateWrapper.kt#L126-L127

And we should add a `try{ }catch` to handle the `NoSuchFieldException` exception.

Doing that uncovers another error, this time related to android.view.ViewStructure

<image src="https://github.com/expo/expo/assets/11707729/b4bab224-2198-4d30-a2dc-988b74e783ca" height="400"/>

@ide Kudo will investigate this (https://linear.app/expo/issue/ENG-10366/[sdk-49]-[android]-[expo-dev-client]-crash-on-android-api-22)

**Comment by ide** [comment]
Thanks. Kudo mentioned he wants to look more closely into how the behavior of `accessFlags` differs between Android versions, which will be helpful, especially for the supported API range of 23+ moving forward.

**Comment by GabrielDierks** [comment]
Is there any estimate when or if this is going to be fixed for the expo-dev-client? 🙏

**Comment by Kudo** [comment]
@GabrielDierks sorry for being late in response. i tried to open a pr to fix the issue and it would only backport to sdk 49 (since from sdk-50, we will drop android api 21-22).

do you have android 22 on hand? would be appreciated if you could help to verify whether the pr works before we publishing the fix? you could get the diff at https://github.com/expo/expo/pull/25248.diff

**Comment by GabrielDierks** [comment]
Tested the diff with a Pixel_XL_API_22, works to debug without crashing! Still had to apply the config-plugin for flipper though that I mentioned in the issue. Thank you so much for taking the time to fix this! May I ask why android api 21-22 are dropped in SDK-50?

**Comment by GabrielDierks** [comment]
Tested the diff with a Pixel_XL_API_22, works to debug without crashing! Still had to apply the config-plugin for flipper though that I mentioned in the issue. Thank you so much for taking the time to fix this! May I ask why android api 21-22 are dropped in SDK-50?

**Comment by Kudo** [comment]
thanks for helping to check the fix. we will proceed the publish the fix.

for flipper, the issue should relate to https://github.com/facebook/flipper/issues/3572. i am sorry that we don't have a better solution in the meantime. disabling flipper as you mentioned sounds to be a good solution in the meantime. from react-native 0.74, flipper would be fully phrased out and it should work fine then.

regarding for dropping android api 21-22, it is because the api 21-22 usage is lower than 2% (according to android studio statistics). also a new android studio created project, it will have min sdk 24. react-native would also drop api 21-22 from 0.74.

**Comment by GabrielDierks** [comment]
Thank you sir for the insights!

**Comment by Kudo** [comment]
@GabrielDierks we published `expo@49.0.17` and `expo-dev-client@2.4.12` with the fix. you should be able to just update packages and remove the patches. thanks again for letting us know the problem and helping to check my fix.

**Comment by iampato** [comment]
@Kudo 
I have the same issue on:
`"expo": "~52.0.27"` & `"expo-dev-client": "~5.0.12"`
Below are the adb logcat logs:
```

03-10 10:56:22.710  1574  3501 I ActivityTaskManager: START u0 {cmp=com.example.sales.app/expo.modules.devlauncher.launcher.errors.DevLauncherErrorActivity} with LAUNCH_MULTIPLE from uid 10411 (BAL_ALLOW_VISIBLE_WINDOW) result code=0
03-10 10:56:22.710  2040  2061 V WindowManagerShell: Transition requested (#214): android.os.BinderProxy@3a430dc TransitionRequestInfo { type = OPEN, triggerTask = TaskInfo{userId=0 taskId=1823 effectiveUid=10411 displayId=0 isRunning=true baseIntent=Intent { act=android.intent.action.VIEW dat=exp+agent-app://expo-development-client/... flg=0x10000000 cmp=com.example.sales.app/.MainActivity } baseActivity=ComponentInfo{com.example.sales.app/com.example.sales.app.MainActivity} topActivity=ComponentInfo{com.example.sales.app/expo.modules.devlauncher.launcher.errors.DevLauncherErrorActivity} origActivity=null realActivity=ComponentInfo{com.example.sales.app/com.example.sales.app.MainActivity} numActivities=2 lastActiveTime=14175452 supportsMultiWindow=true resizeMode=1 isResizeable=true minWidth=-1 minHeight=-1 defaultMinSize=220 token=WCT{android.window.IWindowContainerToken$Stub$Proxy@9f405e5} topActivityType=1 pictureInPictureParams=null shouldDockBigOverlays=false launchIntoPipHostTaskId=-1 lastParentTaskIdBeforePip=-1 displayCutoutSafeInsets=Rect(0, 0 - 0, 0) topActivityInfo=ActivityInfo{ad078ba expo.modules.devlauncher.launcher.errors.DevLauncherErrorActivity} launchCookies=[] positionInParent=Point(0, 0) parentTaskId=-1 isFocused=true isVisible=true isVisibleRequested=true isSleeping=false locusId=null displayAreaFeatureId=1 isTopActivityTransparent=false isTopActivityStyleFloating=false lastNonFullscreenBounds=Rect(351, 828 - 1089, 2268) capturedLink=null capturedLinkTimestamp=0 appCompatTaskInfo=AppCompatTaskInfo { topActivityInSizeCompat=false eligibleForLetterboxEducation= false isLetterboxEducationEnabled= false isLetterboxDoubleTapEnabled= false eligibleForUserAspectRatioButton= false topActivityBoundsLetterboxed= false isFromLetterboxDoubleTap= false topActivityLetterboxVerticalPosition= -1 topActivityLetterboxHorizontalPosition= -1 topActivityLetterboxWidth=1440 topActivityLetterboxHeight=3120 topActivityLetterboxAppWidth=1440 topActivityLetterboxAppHeight=2807 isUserFullscreenOverrideEnabled=false isSystemFullscreenOverrideEnabled=false hasMinAspectRatioOverride=false cameraCompatTaskInfo=CameraCompatTaskInfo { freeformCameraCompatMode=inactive}}}, pipTask = null, remoteTransition = null, displayChange = null, flags = 0, debugId = 214 }
03-10 10:56:22.754  1574  3482 D CoreBackPreview: Window{d107ae3 u0 com.example.sales.app/expo.modules.devlauncher.launcher.errors.DevLauncherErrorActivity}: Setting back callback OnBackInvokedCallbackInfo{mCallback=android.window.IOnBackInvokedCallback$Stub$Proxy@ec0be5e, mPriority=0, mIsAnimationCallback=false}
03-10 10:56:22.779  1574  3524 D ConnectivityService: requestNetwork for uid/pid:10411/17628 activeRequest: null callbackRequest: 929 [NetworkRequest [ REQUEST id=930, [ Capabilities: INTERNET&NOT_RESTRICTED&TRUSTED&NOT_VCN_MANAGED&NOT_BANDWIDTH_CONSTRAINED Uid: 10411 RequestorUid: 10411 RequestorPkg: com.example.sales.app UnderlyingNetworks: Null] ]] callback flags: 0 order: 2147483647 isUidTracked: false declaredMethods: AVAIL|LOSING|LOST|UNAVAIL|NC|LP
03-10 10:56:22.780  1574  1646 I ActivityTaskManager: Displayed com.example.sales.app/expo.modules.devlauncher.launcher.errors.DevLauncherErrorActivity for user 0: +77ms
03-10 10:56:22.780  2040  2061 V WindowManagerShell: onTransitionReady (#214) android.os.BinderProxy@3a430dc: {id=214 t=OPEN f=0x0 trk=0 r=[0@Point(0, 0)] c=[{null m=OPEN f=FILLS_TASK leash=Surface(name=ActivityRecord{98e1147 u0 com.example.sales.app/expo.modules.devlauncher.launcher.errors.DevLauncherErrorActivity)/@0x968436b sb=Rect(0, 0 - 0, 0) eb=Rect(0, 0 - 1440, 3120) epz=Point(1440, 3120) d=-1->0 r=-1->0:-1 bc=fffafafa component=com.example.sales.app/expo.modules.devlauncher.launcher.errors.DevLauncherErrorActivity opt={t=FROM_STYLE mUserId=0}},{null m=TO_BACK f=FILLS_TASK leash=Surface(name=ActivityRecord{e01abf u0 com.example.sales.app/.MainActivity)/@0x5b17ec8 sb=Rect(0, 0 - 1440, 3120) eb=Rect(0, 0 - 1440, 3120) epz=Point(1440, 3120) d=0 bc=fffafafa component=com.example.sales.app/.MainActivity opt={t=FROM_STYLE mUserId=0}}]}
03-10 10:56:22.781  2040  2061 V WindowManagerShell: Transition doesn't have explicit remote, search filters for match for {id=214 t=OPEN f=0x0 trk=0 r=[0@Point(0, 0)] c=[{null m=OPEN f=FILLS_TASK leash=Surface(name=ActivityRecord{98e1147 u0 com.example.sales.app/expo.modules.devlauncher.launcher.errors.DevLauncherErrorActivity)/@0x968436b sb=Rect(0, 0 - 0, 0) eb=Rect(0, 0 - 1440, 3120) epz=Point(1440, 3120) d=-1->0 r=-1->0:-1 bc=fffafafa component=com.example.sales.app/expo.modules.devlauncher.launcher.errors.DevLauncherErrorActivity opt={t=FROM_STYLE mUserId=0}},{null m=TO_BACK f=FILLS_TASK leash=Surface(name=ActivityRecord{e01abf u0 com.example.sales.app/.MainActivity)/@0x5b17ec8 sb=Rect(0, 0 - 1440, 3120) eb=Rect(0, 0 - 1440, 3120) epz=Point(1440, 3120) d=0 bc=fffafafa component=com.example.sales.app/.MainActivity opt={t=FROM_STYLE mUserId=0}}]}
03-10 10:56:22.781  2040  2061 V WindowManagerShell: start default transition animation, info = {id=214 t=OPEN f=0x0 trk=0 r=[0@Point(0, 0)] c=[{null m=OPEN f=FILLS_TASK leash=Surface(name=ActivityRecord{98e1147 u0 com.example.sales.app/expo.modules.devlauncher.launcher.errors.DevLauncherErrorActivity)/@0x968436b sb=Rect(0, 0 - 0, 0) eb=Rect(0, 0 - 1440, 3120) epz=Point(1440, 3120) d=-1->0 r=-1->0:-1 bc=fffafafa component=com.example.sales.app/expo.modules.devlauncher.launcher.errors.DevLauncherErrorActivity opt={t=FROM_STYLE mUserId=0}},{null m=TO_BACK f=FILLS_TASK leash=Surface(name=ActivityRecord{e01abf u0 com.example.sales.app/.MainActivity)/@0x5b17ec8 sb=Rect(0, 0 - 1440, 3120) eb=Rect(0, 0 - 1440, 3120) epz=Point(1440, 3120) d=0 bc=fffafafa component=com.example.sales.app/.MainActivity opt={t=FROM_STYLE mUserId=0}}]}
03-10 10:56:22.782  1574  1901 D OemPaidWifiNetworkFactory: got request NetworkRequest [ REQUEST id=930, [ Capabilities: INTERNET&NOT_RESTRICTED&TRUSTED&NOT_VCN_MANAGED&NOT_BANDWIDTH_CONSTRAINED Uid: 10411 RequestorUid: 10411 RequestorPkg: com.example.sales.app UnderlyingNetworks: Null] ]
03-10 10:56:22.782  1574  1901 D MultiInternetWifiNetworkFactory: got request NetworkRequest [ REQUEST id=930, [ Capabilities: INTERNET&NOT_RESTRICTED&TRUSTED&NOT_VCN_MANAGED&NOT_BANDWIDTH_CONSTRAINED Uid: 10411 RequestorUid: 10411 RequestorPkg: com.example.sales.app UnderlyingNetworks: Null] ]
03-10 10:56:22.783  1574  1646 V WindowManager: Sent Transition (#214) createdAt=03-10 10:56:22.703 via request=TransitionRequestInfo { type = OPEN, triggerTask = TaskInfo{userId=0 taskId=1823 effectiveUid=10411 displayId=0 isRunning=true baseIntent=Intent { act=android.intent.action.VIEW dat=exp+agent-app://expo-development-client/... flg=0x10000000 cmp=com.example.sales.app/.MainActivity } baseActivity=ComponentInfo{com.example.sales.app/com.example.sales.app.MainActivity} topActivity=ComponentInfo{com.example.sales.app/expo.modules.devlauncher.launcher.errors.DevLauncherErrorActivity} origActivity=null realActivity=ComponentInfo{com.example.sales.app/com.example.sales.app.MainActivity} numActivities=2 lastActiveTime=14175452 supportsMultiWindow=true resizeMode=1 isResizeable=true minWidth=-1 minHeight=-1 defaultMinSize=220 token=WCT{RemoteToken{922379b Task{cf78b16 #1823 type=standard A=10411:com.example.sales.app isResizeable=true minWidth=-1 minHeight=-1}}} topActivityType=1 pictureInPictureParams=null shouldDockBigOverlays=false launchIntoPipHostTaskId=-1 lastParentTaskIdBeforePip=-1 displayCutoutSafeInsets=Rect(0, 0 - 0, 0) topActivityInfo=ActivityInfo{99f94d1 expo.modules.devlauncher.launcher.errors.DevLauncherErrorActivity} launchCookies=[] positionInParent=Point(0, 0) parentTaskId=-1 isFocused=true isVisible=true isVisibleRequested=true isSleeping=false locusId=null displayAreaFeatureId=1 isTopActivityTransparent=false isTopActivityStyleFloating=false lastNonFullscreenBounds=Rect(351, 828 - 1089, 2268) capturedLink=null capturedLinkTimestamp=0 appCompatTaskInfo=AppCompatTaskInfo { topActivityInSizeCompat=false eligibleForLetterboxEducation= false isLetterboxEducationEnabled= false isLetterboxDoubleTapEnabled= false eligibleForUserAspectRatioButton= false topActivityBoundsLetterboxed= false isFromLetterboxDoubleTap= false topActivityLetterboxVerticalPosition= -1 topActivityLetterboxHorizontalPosition= -1 topActivityLetterboxWidth=1440 topActivityLetterboxHeight=3120 topActivityLetterboxAppWidth=1440 topActivityLetterboxAppHeight=2807 isUserFullscreenOverrideEnabled=false isSystemFullscreenOverrideEnabled=false hasMinAspectRatioOverride=false cameraCompatTaskInfo=CameraCompatTaskInfo { freeformCameraCompatMode=inactive}}}, pipTask = null, remoteTransition = null, displayChange = null, flags = 0, debugId = 214 }
03-10 10:56:22.783  1574  1646 V WindowManager:         {null m=OPEN f=FILLS_TASK leash=Surface(name=ActivityRecord{98e1147 u0 com.example.sales.app/expo.modules.devlauncher.launcher.errors.DevLauncherErrorActivity)/@0xa40635b sb=Rect(0, 0 - 0, 0) eb=Rect(0, 0 - 1440, 3120) epz=Point(1440, 3120) d=-1->0 r=-1->0:-1 bc=fffafafa component=com.example.sales.app/expo.modules.devlauncher.launcher.errors.DevLauncherErrorActivity opt={t=FROM_STYLE mUserId=0}},
03-10 10:56:22.783  1574  1646 V WindowManager:         {null m=TO_BACK f=FILLS_TASK leash=Surface(name=ActivityRecord{e01abf u0 com.example.sales.app/.MainActivity)/@0xbd33a24 sb=Rect(0, 0 - 1440, 3120) eb=Rect(0, 0 - 1440, 3120) epz=Point(1440, 3120) d=0 bc=fffafafa component=com.example.sales.app/.MainActivity opt={t=FROM_STYLE mUserId=0}}
03-10 10:56:22.789  1574  1901 D WifiNetworkFactory: got request NetworkRequest [ REQUEST id=930, [ Capabilities: INTERNET&NOT_RESTRICTED&TRUSTED&NOT_VCN_MANAGED&NOT_BANDWIDTH_CONSTRAINED Uid: 10411 RequestorUid: 10411 RequestorPkg: com.example.sales.app UnderlyingNetworks: Null] ]
03-10 10:56:22.789  1574  1901 D UntrustedWifiNetworkFactory: got request NetworkRequest [ REQUEST id=930, [ Capabilities: INTERNET&NOT_RESTRICTED&TRUSTED&NOT_VCN_MANAGED&NOT_BANDWIDTH_CONSTRAINED Uid: 10411 RequestorUid: 10411 RequestorPkg: com.example.sales.app UnderlyingNetworks: Null] ]
03-10 10:56:22.911  1574  1870 I ImeTracker: com.example.sales.app:97dffaf8: onRequestHide at ORIGIN_SERVER reason HIDE_UNSPECIFIED_WINDOW fromUser false
03-10 10:56:22.911  1574  1870 I ImeTracker: com.example.sales.app:97dffaf8: onCancelled at PHASE_SERVER_SHOULD_HIDE
03-10 10:56:22.913  3206  3206 I GoogleInputMethodService: GoogleInputMethodService.onStartInput():1568 onStartInput(EditorInfo{EditorInfo{packageName=com.example.sales.app, inputType=0, inputTypeString=NULL, enableLearning=false, autoCorrection=false, autoComplete=false, imeOptions=0, privateImeOptions=null, actionName=UNSPECIFIED, actionLabel=null, initialSelStart=-1, initialSelEnd=-1, initialCapsMode=0, label=null, fieldId=0, fieldName=null, extras=null, hintText=null, hintLocales=[]}}, false)
03-10 10:56:22.913  1574  3524 W PackageConfigPersister: App-specific configuration not found for packageName: com.example.sales.app and userId: 0
```
